sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala

```diff
@@ -177,4 +177,18 @@ class CanonicalizeSuite extends SparkFunSuite {
     assert(expr.semanticEquals(attr))
     assert(attr.semanticEquals(expr))
   }
+
+  test("SPARK-38030: Canonicalize Cast should remove nullability of target dataType") {
+    val structType = StructType(Seq(StructField("name", StringType, nullable = false)))
+    val attr = AttributeReference("col", structType)()
+    for (cast <- Seq(
+        Cast(attr, structType),
+        AnsiCast(attr, structType),
+        TryCast(attr, structType))) {
```

Comment:
```suggestion
    Seq(Cast.apply _, AnsiCast.apply _, TryCast.apply _)
        .map(_.apply(attr, structType, None))
        .foreach { cast =>
```
Not sure if this is better, but it does deduplicate the args. A bit messy either way.

Reply:
The existing approach is simpler to read IMO, but open to changing it if more folks think the second approach is better.